quaddtype/numpy_quaddtype/src/casts.cpp

```diff
@@ -26,88 +27,116 @@ extern "C" {
 }
 
-static int
-quad_to_quad_strided_loop_unaligned(PyArrayMethod_Context *context, char *const data[],
-                                    npy_intp const dimensions[], npy_intp const strides[],
-                                    void *NPY_UNUSED(auxdata))
+// Helper function for quad-to-quad same_value check (inter-backend)
+// NOTE: the inter-backend uses `double` as intermediate,
+// so only values that can be exactly represented in double can pass same_value check
+static inline int
+quad_to_quad_same_value_check(const quad_value *in_val, QuadBackendType backend_in,
+                              const quad_value *out_val, QuadBackendType backend_out)
 {
-    npy_intp N = dimensions[0];
-    char *in_ptr = data[0];
-    char *out_ptr = data[1];
-    npy_intp in_stride = strides[0];
-    npy_intp out_stride = strides[1];
-
-    QuadPrecDTypeObject *descr_in = (QuadPrecDTypeObject *)context->descriptors[0];
-    QuadPrecDTypeObject *descr_out = (QuadPrecDTypeObject *)context->descriptors[1];
-
-    // inter-backend casting
-    if (descr_in->backend != descr_out->backend) {
-        while (N--) {
-            quad_value in_val, out_val;
-            if (descr_in->backend == BACKEND_SLEEF) {
-                memcpy(&in_val.sleef_value, in_ptr, sizeof(Sleef_quad));
-                out_val.longdouble_value = Sleef_cast_to_doubleq1(in_val.sleef_value);
-            }
-            else {
-                memcpy(&in_val.longdouble_value, in_ptr, sizeof(long double));
-                out_val.sleef_value = Sleef_cast_from_doubleq1(in_val.longdouble_value);
-            }
-            memcpy(out_ptr, &out_val,
-                   (descr_out->backend == BACKEND_SLEEF) ? sizeof(Sleef_quad)
-                                                         : sizeof(long double));
-            in_ptr += in_stride;
-            out_ptr += out_stride;
+    // Convert output back to input backend for comparison
+    quad_value roundtrip;
+    
+    if (backend_in == BACKEND_SLEEF) {
+        // Input was SLEEF, output is longdouble
+        // Convert longdouble back to SLEEF for comparison
+        long double ld = out_val->longdouble_value;
+        if (std::isnan(ld)) {
+            roundtrip.sleef_value = QUAD_PRECISION_NAN;
         }
-
-        return 0;
+        else if (std::isinf(ld)) {
+            roundtrip.sleef_value = (ld > 0) ? QUAD_PRECISION_INF : QUAD_PRECISION_NINF;
+        }
+        else {
+            Sleef_quad temp = Sleef_cast_from_doubleq1(static_cast<double>(ld));
+            memcpy(&roundtrip.sleef_value, &temp, sizeof(Sleef_quad));
+        }
+        
+        // Compare in SLEEF domain
+        if (Sleef_iunordq1(in_val->sleef_value, roundtrip.sleef_value))
+            return 1;  // Both NaN
+        if (Sleef_icmpeqq1(in_val->sleef_value, roundtrip.sleef_value))
+            return 1;  // Equal
+        if (Sleef_icmpeqq1(in_val->sleef_value, QUAD_ZERO) && Sleef_icmpeqq1(roundtrip.sleef_value, QUAD_ZERO))
```

Comment:
Why do we have this extra check? If the previous equality check doesn't catch +0.0 == -0.0, then this one would still check (+0.0 == +0.0) && (-0.0 == +0.0) which is no better